sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -764,35 +865,113 @@ public Write withoutValidation() {
     }
 
     /**
-     * Returns a new {@link BigtableIO.Write} that will write using the given Cloud Bigtable service
-     * implementation.
+     * Returns a new {@link BigtableIO.Write} that will use an official Bigtable emulator.
      *
      * <p>This is used for testing.
+     */
+    @VisibleForTesting
+    public Write withEmulator(String emulatorHost) {
+      BigtableConfig config = getBigtableConfig();
+      return toBuilder().setBigtableConfig(config.withEmulator(emulatorHost)).build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with the attempt timeout. Attempt timeout controls the
+     * timeout for each remote call.
      *
      * <p>Does not modify this object.
      */
-    Write withBigtableService(BigtableService bigtableService) {
-      BigtableConfig config = getBigtableConfig();
-      return toBuilder().setBigtableConfig(config.withBigtableService(bigtableService)).build();
+    public Write withAttemptTimeout(Duration timeout) {
+      checkArgument(timeout.isLongerThan(Duration.ZERO), "attempt timeout must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setAttemptTimeout(timeout).build())
+          .build();
     }
 
     /**
-     * Returns a new {@link BigtableIO.Write} that will use an official Bigtable emulator.
+     * Returns a new {@link BigtableIO.Write} with the operation timeout. Operation timeout has
+     * ultimate control over how long the logic should keep trying the remote call until it gives up
+     * completely.
      *
-     * <p>This is used for testing.
+     * <p>Does not modify this object.
      */
+    public Write withOperationTimeout(Duration timeout) {
+      checkArgument(timeout.isLongerThan(Duration.ZERO), "operation timeout must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setOperationTimeout(timeout).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with batch element size. After this many elements are
+     * accumulated, they will be wrapped up in a batch and sent to Bigtable.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withBatchElements(long size) {
+      checkArgument(size > 0, "batch element size must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setBatchElements(size).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with batch bytes size. After this many bytes are
+     * accumulated, the elements will be wrapped up in a batch and sent to Bigtable.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withBatchBytes(long size) {
```

Comment:
nit, can we rename this to be a bit more precise? maybe withMaxBytesPerBatch? or batchByteThreshold?

Reply:
Also make sure to update the populateDisplay data as well